controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,361 @@
+	version, err := contract.ControlPlane().Version().Get(cp)
+	if err != nil {
+		log.Infof("We failed on getting control plane version %v %v", err, cp)
+		return false
+	}
+	if *version != cluster.Spec.Topology.Version {
+		log.Infof("We failed on comparing control plane version %v", cp)
+		return false
+	}
+	// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+	if cluster.Spec.Topology.ControlPlane.Replicas != nil {
+		replicas, err := contract.ControlPlane().Replicas().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane replicas %v", cp)
+			return false
+		}
+		// Check for Control Plane replicase if it's set in the Cluster.Spec.Topology
+		if int32(*replicas) != *cluster.Spec.Topology.ControlPlane.Replicas {
+			log.Infof("We failed on comparing control plane replicas %v", cp)
+			return false
+		}
+	}
+	log.Infof("ControlPlane passed with %v", cp)
+	if clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		cpInfra, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(cp)
+		if err != nil {
+			log.Infof("We failed on getting control plane infrastructure %v,  %v", err, cp)
+			return false
+		}
+		if ok := referenceExistsWithCorrectKindAndAPIVersion(cpInfra,
+			builder.GenericInfrastructureMachineTemplateKind,
+			builder.InfrastructureGroupVersion); !ok {
+			log.Infof("Kind or version wrong from InfrastructureMachineTemplate %v", cp)
+			return false
+		}
+		if _, ok := getAndEnsureLabelsAndAnnotations(*cpInfra, cluster.Name); !ok {
+			return false
+		}
+		log.Infof("ControlPlaneInfrastructure passed with %v", cpInfra)
+	}
+	return true
+}
+
+// assertMachineDeploymentsReconcile checks if the MachineDeployments:
+// 1) Are created in the correct number.
+// 2) Have the correct labels (TopologyOwned, ClusterName, MachineDeploymentName).
+// 3) Have the correct finalizer applied.
+// 4) Have the correct replicas and version.
+// 6) Have the correct Kind/APIVersion and Labels/Annotations for BoostrapRef and InfrastructureRef templates.
+func assertMachineDeploymentsReconcile(cluster *clusterv1.Cluster) bool {
+	// List all created machine deployments to ensure the expected numbers are c*reated.
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	if err := env.List(ctx, machineDeployments); err != nil {
+		log.Infof("machine deployments not listed %v", err)
+		return false
+	}
+	clusterMDs := make([]clusterv1.MachineDeployment, 0)
+	// Run through all machine deployments and add only those with the TopologyOwnedLable and the correct
+	// ClusterLabelName to the items for further testing.
+	for _, m := range machineDeployments.Items {
+		// If the machineDeployment doesn't have the ClusterTopologyOwnedLabel and the ClusterLabelName ignore.
+		md := m.DeepCopy()
+		if !(ensureClusterTopologyOwnedLabel(md) &&
+			ensureClusterLabelName(md, cluster.Name)) {
+			continue
```

Comment:
I think we should fail here, because this should not happen in this test

Reply:
Sure - I wasn't sure whether to leave some room for this test to be used for different scenarios - there' s a couple of cleanups (like t.Run) I need to do to make it much more like a single run test.

Reply:
I think these utility functions could potentially be used in other integration tests, so the behaviour here is just about checking whether the reconcile happened correctly for whatever input.